src/azul/project/hca/metadata_api/__init__.py

```diff
@@ -0,0 +1,598 @@
+from abc import ABC, abstractmethod
+from types import SimpleNamespace
+from typing import List, MutableMapping, Any, Set, Optional, Union, Iterable
+from uuid import UUID
+
+from dataclasses import dataclass
+
+from azul.dss_bundle import DSSBundle
+from azul.types import AnyJSON, JSON
+from .expando import Expando
+
+
+UUID4 = UUID
+_ObjectifiedJSON = Union[str, int, float, bool, None, MutableMapping[str, Any], List[Any]]
+ObjectifiedJSON = Union[MutableMapping[str, _ObjectifiedJSON], SimpleNamespace]
+
+mydataclass = dataclass(init=False)
+
+
+@dataclass
+class ManifestEntry:
+    content_type: str
+    crc32c: str
+    indexed: bool
+    name: str
+    s3_etag: str
+    sha1: str
+    sha256: str
+    size: int
+    uuid: UUID4
+    version: str
+
+    @classmethod
+    def from_metadata(cls, json: ObjectifiedJSON):
+        json.content_type = json.pop("content-type")
+        # noinspection PyTypeChecker
+        json.uuid = UUID4(json.uuid)
+        # noinspection PyArgumentList
+        return cls(**json)
+
+
+class EntityVisitor(ABC):
+
+    @abstractmethod
+    def visit(self, entity: 'Entity') -> None:
+        raise NotImplementedError()
+
+
+@mydataclass
+class Entity:
+    document_id: UUID4
+    _json: ObjectifiedJSON
+    _source: str
+
+    @classmethod
+    def from_metadata(cls, json: ObjectifiedJSON):
+        described_by = json.content.describedBy.rpartition('/')[2]
+        sub_cls = object_types[described_by]
+        return sub_cls(json)
+
+    def __init__(self, json: ObjectifiedJSON) -> None:
+        super().__init__()
+        self._source = json.content.describedBy.rpartition('/')[2]
```

Comment:
`source` is an odd name for this, I think. Also don't like the duplication with the line from above.